contrib/win32/win32compat/misc.c

```diff
@@ -680,8 +691,30 @@ link(const char *oldpath, const char *newpath)
 int
 w32_rename(const char *old_name, const char *new_name)
 {
-	wchar_t *resolvedOldPathName_utf16 = utf8_to_utf16(sanitized_path(old_name));
-	wchar_t *resolvedNewPathName_utf16 = utf8_to_utf16(sanitized_path(new_name));
+	char old_name_expanded[PATH_MAX] = {0, };
+	char new_name_expanded[PATH_MAX] = {0, };
+
+	// When sshd service runs "ssh-keygen -A" command, the filenames will have SSHDIR.
+	if (NULL != strstr(old_name, SSHDIR)) {
+		strcat_s(old_name_expanded, _countof(old_name_expanded), get_ssh_dir_path());
+
+		// append filename. path - "c:\\ProgramData\\openssh\\<filename>"
+		strcat_s(old_name_expanded, _countof(old_name_expanded), &old_name[strlen(SSHDIR)]);
+	} else {
+		strcpy_s(old_name_expanded, _countof(old_name_expanded), old_name);
+	}
+
+	if (NULL != strstr(new_name, SSHDIR)) {
+		strcat_s(new_name_expanded, _countof(new_name_expanded), get_ssh_dir_path());
+
+		// append filename. path - "c:\\ProgramData\\openssh\\<filename>"
+		strcat_s(new_name_expanded, _countof(new_name_expanded), &new_name[strlen(SSHDIR)]);
+	} else {
+		strcpy_s(new_name_expanded, _countof(new_name_expanded), new_name);
```

Comment:
Looks like all of this can be simplified but just resolving __PROGRAMDATA_ within sanitize_path ?